sw/device/tests/spi_device_flash_smoketest.c

```diff
@@ -0,0 +1,166 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/status.h"
+#include "sw/device/lib/dif/dif_spi_device.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/spi_device_testutils.h"
+#include "sw/device/lib/testing/spi_flash_emulator.h"
+#include "sw/device/lib/testing/spi_flash_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG(.enable_uart_flow_control = true);
+
+enum {
+  kSpiDeviceDatasetSize = 128,
+  kSpiDeviceFlashAddress = 0x00,
+};
+
+// A set of bytes to be send out by spi_device.
+const uint8_t kSpiTxData[kSpiDeviceDatasetSize] = {
```

Comment:
This test seems fine as a smoke test, but for testing the "LAST_READ_ADDR" feature and eFlash reads, it doesn't really reflect the use case or provide great coverage.

An eFlash-specific test would probably do something like this:

1. Host writes (via ujson MemWrite or SPI flash PAGE_PROGRAM) at least 4 KiB of data in an array that the device would stream out through the eFlash read buffer.
2. Host tells the DUT (via MemWrite32) to begin streaming the data.
3. DUT copies data to the spi_device read buffer, and signals the host that it is ready.
4. Host begins continuously reading a random amount of data (but larger than 1 KiB) via SPI, while the DUT feeds the next half of the buffer. DUT monitors the flip or watermark events to determine whether it can begin writing the next half.
5. Host checks data correctness of what it read + the LAST_READ_ADDR (can probe LAST_READ_ADDR via a ujson MemRead32 or something).

Reply:
Agreed, will put that in an issue and merge this test for now. Thanks.

Reply:
Issue https://github.com/lowRISC/opentitan/issues/21380